pilot/helpers/agents/CodeMonkey.py

```diff
@@ -1,38 +1,353 @@
-from const.function_calls import GET_FILES, IMPLEMENT_CHANGES
+import os.path
+import re
+from typing import Optional
+
 from helpers.AgentConvo import AgentConvo
 from helpers.Agent import Agent
+from helpers.files import get_file_contents
+from logger.logger import logger
 
 
 class CodeMonkey(Agent):
+    save_dev_steps = True
+
     def __init__(self, project, developer):
         super().__init__('code_monkey', project)
         self.developer = developer
 
-    def implement_code_changes(self, convo, task_description, code_changes_description, step, step_index=0):
-        if convo is None:
+    def implement_code_changes(
+        self,
+        convo: Optional[AgentConvo],
+        task_description: str,
+        code_changes_description: str,
+        step: dict[str, str],
+        step_index: int = 0,
+    ) -> AgentConvo:
+        """
+        Implement code changes described in `code_changes_description`.
+
+        :param convo: AgentConvo instance (optional)
+        :param task_description: description of the task
+        :param code_changes_description: description of the code changes
+        :param step: information about the step being implemented
+        :param step_index: index of the step to implement
+        """
+        standalone = False
+        if not convo:
+            standalone = True
             convo = AgentConvo(self)
 
-        # files_needed = convo.send_message('development/task/request_files_for_code_changes.prompt', {
-        #     "step_description": code_changes_description,
-        #     "directory_tree": self.project.get_directory_tree(True),
-        #     "step_index": step_index,
-        #     "finished_steps": ', '.join(f"#{j}" for j in range(step_index))
-        # }, GET_FILES)
+        files = self.project.get_all_coded_files()
+
+        # If we're called as a result of debugging, we don't have the name/path of the file
+        # to modify so we need to figure that out first.
+        if 'path' not in step or 'name' not in step:
+            files_to_change = self.identify_files_to_change(code_changes_description, files)
+
+            if len(files_to_change) == 0:
+                # This should never happen, but let's not crash if it does
+                logger.warning(f"No relevant files found for code change: {code_changes_description}")
+                return convo
+
+            # In practice, we always have exactly one file to change, so we can just pick the first in
```

Comment:
Old prompts were quite ambiguous whether it can be one or more files so when I started writing that code I wasn't sure what needed to be done. In all my testing I've never had this handling more than one file. I also tweaked the prompt to ask for one file, not multiple.

I'll be changing the prompt to use JSON call so we'll force the LLM to always spew out exactly one.

Reply:
We now force exactly one file through JSON response in 8e301fa1cee720bd789e977e630c8adf55676bb1